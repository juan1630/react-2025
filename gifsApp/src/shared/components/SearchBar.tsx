import { useEffect, useState } from "react";

interface Props {
  placeholder?: string;
  onQuery: (query: string) => void;
}
export const CustomSearch = ({ placeholder, onQuery }: Props) => {
  const [query, setQuery] = useState("");
  useEffect(() => {
    
    const timeoutId = setTimeout(()=> {
      onQuery(query)
    } , 700)

    return ()=> {
      clearTimeout(timeoutId)
    }
  }, [query, setQuery])

  const handleSearch = () => {
    onQuery(query);
    setQuery('')
  };

  const handleKeyDown = (event: React.KeyboardEvent<HTMLInputElement>) => {
    if (event.key === "Enter") {
      handleSearch();
    }
  };

  return (
    <div className="search-container">
      <input
        type="text"
        name=""
        value={query}
        onChange={(event) => setQuery(event.target.value)}
        placeholder={placeholder ? placeholder : "Buscar"}
        onKeyDown={handleKeyDown}
      />
      <button onClick={handleSearch}>Buscar</button>
    </div>
  );
};
