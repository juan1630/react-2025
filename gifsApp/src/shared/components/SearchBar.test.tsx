import { describe, expect, test, vi } from "vitest";
import { fireEvent, render, screen, waitFor } from "@testing-library/react";
import { CustomSearch } from "./SearchBar";

describe("should render Searchbar correct", () => {
  test("should render searchbar correctly", () => {
    const { container } = render(<CustomSearch onQuery={() => {}} />);
    expect(container).toMatchSnapshot();
    expect(screen.getByRole("textbox")).toBeDefined();
  });

  test("should call onQuery with the correct value after 700ms", async () => {
    const onQuery = vi.fn();
    render(<CustomSearch onQuery={onQuery} />);

    const input = screen.getByRole("textbox");
    fireEvent.change(input, { target: { value: "test" } });

    // await new Promise((resolve) => setTimeout(resolve, 700))
    await waitFor(() => {
      expect(onQuery).toHaveBeenCalled();
      expect(onQuery).toHaveBeenCalledWith("test");
    });
  });

  test("should call only once with the last value (debounce)", async () => {
    const onQuery = vi.fn();
    render(<CustomSearch onQuery={onQuery} />);

    const input = screen.getByRole("textbox");
    fireEvent.change(input, { target: { value: "t" } });
    fireEvent.change(input, { target: { value: "te" } });

    await waitFor(() => {
      expect(onQuery).toHaveBeenCalledTimes(1);
      expect(onQuery).toHaveBeenCalledWith("te");
    });
  });

  test("should call onquery when button is clicked with the input value", () => {
    const onQuery = vi.fn();
    render(<CustomSearch onQuery={onQuery} />);

    const input = screen.getByRole("textbox");
    fireEvent.change(input, { target: { value: "t" } });

    const button = screen.getByRole("button");
    fireEvent.click(button);

    expect(onQuery).toHaveBeenCalledTimes(1);
    expect(onQuery).toHaveBeenCalledWith("t");
  });

  test('should the input has the correct placeholder value', () => {
    
    const placeholder = 'Buscar'
    render(<CustomSearch onQuery={()=> {}}  placeholder={placeholder}/>);
    
    expect(screen.getByPlaceholderText(placeholder)).toBeDefined()

  })
});
