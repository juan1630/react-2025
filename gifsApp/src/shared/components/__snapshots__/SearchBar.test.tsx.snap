// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`should render Searchbar correct > should render searchbar correctly 1`] = `
<div>
  <div
    class="search-container"
  >
    <input
      name=""
      placeholder="Buscar"
      type="text"
      value=""
    />
    <button>
      Buscar
    </button>
  </div>
</div>
`;
